python/pyarrow/io.pxi

```diff
@@ -101,28 +101,54 @@ cdef class NativeFile:
         """
         raise UnsupportedOperation()
 
+    @property
+    def closed(self):
+        if self.is_readable:
+            return self.input_stream.get().closed()
+        elif self.is_writable:
+            return self.output_stream.get().closed()
+        else:
+            return True
+
     def close(self):
         if not self.closed:
             with nogil:
                 if self.is_readable:
-                    check_status(self.rd_file.get().Close())
+                    check_status(self.input_stream.get().Close())
                 else:
-                    check_status(self.wr_file.get().Close())
-        self.closed = True
+                    check_status(self.output_stream.get().Close())
 
     def flush(self):
         """Flush the buffer stream, if applicable.
 
         No-op to match the IOBase interface."""
         self._assert_open()
 
-    cdef read_handle(self, shared_ptr[RandomAccessFile]* file):
+    cdef set_random_access_file(self, shared_ptr[RandomAccessFile] handle):
+        self.input_stream = <shared_ptr[InputStream]> handle
+        self.random_access = handle
+        self.is_seekable = True
+
+    cdef set_input_stream(self, shared_ptr[InputStream] handle):
+        self.input_stream = handle
+        self.random_access.reset()
+        self.is_seekable = False
+
+    cdef set_output_stream(self, shared_ptr[OutputStream] handle):
+        self.output_stream = handle
```

Comment:
seekable should get set to False here? (maybe esoteric)

Reply:
IIRC we have read-write streams somewhere...